compiler/lib/macro.ml

```diff
@@ -0,0 +1,80 @@
+(* Js_of_ocaml compiler
+ * http://www.ocsigen.org/js_of_ocaml/
+ * Copyright (C) 2019 Ty Overby
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, with linking exception;
+ * either version 2.1 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
+ *)
+
+open Stdlib
+module J = Javascript
+
+let zero = J.ENum "0"
+
+let one = J.ENum "1"
+
+let make_block ~tag ~args =
+  J.EArr (List.map ~f:(fun x -> Some x) (J.ENum (string_of_int tag) :: args))
+
+let tag e = J.EAccess (e, zero)
+
+let length e =
+  let underlying = J.EDot (e, "length") in
+  J.EBin (J.Minus, underlying, one)
+
+(* JavaScript engines recognize the pattern
+   'typeof x==="number"'; if the string is shared,
+   less efficient code is generated. *)
+
+let is_block e = J.EBin (J.NotEqEq, J.EUn (J.Typeof, e), J.EStr ("number", `Bytes))
+
+let is_immediate e = J.EBin (J.EqEqEq, J.EUn (J.Typeof, e), J.EStr ("number", `Bytes))
+
+let field e i =
+  match i with
+  | J.ENum n ->
+      let idx = int_of_string n in
+      let adjusted = J.ENum (string_of_int (idx + 1)) in
+      J.EAccess (e, adjusted)
+  | J.EUn (J.Neg, _) -> failwith "Negative field indexes are not allowed"
+  | _ ->
+      let adjusted = J.EBin (J.Plus, one, i) in
+      J.EAccess (e, adjusted)
+
+class macro_mapper =
+  object (m)
+    inherit Js_traverse.map as super
+
+    method expression x =
+      let module J = Javascript in
+      match x with
+      | J.ECall (J.EVar (J.S {name; _}), args, _) -> (
+        match name, args with
+        | "BLOCK", J.ENum tag :: (_ :: _ as args) ->
+            make_block
+              ~tag:(int_of_string tag)
```

Comment:
We should validate the string (will be incorrect with the JavaScript octal notation)